utils/similarity/encode.py

```diff
@@ -0,0 +1,188 @@
+import os
+import sys
+
+from slither import Slither
+from slither.slithir.operations import *
+from slither.slithir.variables import *
+from slither.core.declarations import *
+from slither.solc_parsing.declarations.function import *
+from slither.core.solidity_types import *
+
+from slither.solc_parsing.variables.state_variable import *
+from slither.solc_parsing.variables.local_variable import *
+from slither.solc_parsing.variables.local_variable_init_from_tuple import *
+
+logger = logging.getLogger("Slither-simil")
+
+def load_contracts(dirname, ext=None, nsamples=None):
+    r = []
+    walk = list(os.walk(dirname))
+    for x, y, files in walk:
+        for f in files: 
+            if ext is None or f.endswith(ext):
+                r.append(x + "/".join(y) + "/" + f)
+
+    if nsamples is None:
+        return r
+    else:
+        # TODO: shuffle
+        return r[:nsamples]
+
+def ntype(_type):
+    if isinstance(_type, ElementaryType):
+        _type = str(_type)
+    elif isinstance(_type, ArrayType):
+        if isinstance(_type.type, ElementaryType):
+            _type = str(_type)
+        else:
+            _type = "user_defined_array"
+    elif isinstance(_type, Structure):
+        print(_type)
+        _type = str(_type)
+    elif isinstance(_type, Enum):
+        print(_type)
+        _type = str(_type)
+    elif isinstance(_type, MappingType):
+        _type = str(_type)
+    elif isinstance(_type, UserDefinedType):
+        _type = "user_defined_type"  # TODO: this could be Contract, Enum or Struct
+    else:
+        _type = str(_type)
+
+    _type = _type.replace(" memory","")
+    _type = _type.replace(" storage ref","")
+
+    if "struct" in _type:
+        return "struct"
+    elif "enum" in _type:
+        return "enum"
+    elif "tuple" in _type:
+        return "tuple"
+    elif "contract" in _type:
+        return "contract"
+    elif "mapping" in _type:
+        return "mapping"
+    else:
+        return _type.replace(" ","_")
+
+def encode_ir(ir):
```

Comment:
I would split this into two functions `encode_ir` and `encode_value`. 
SlithIR is flat, so all the `ir.value` , `ir.rvalue` will never return an Operation